### PR TITLE
fix type instability in binomial

### PR DESCRIPTION
**Pull request by oscardssmith**

### CODE REVIEW

**Comment by simeonschaub** [comment]
Could you add a test for this? Would also be good to check if this still works for `Bool`s.

**Comment by oscardssmith** [comment]
It's hard to test this since the output was previously being converted so the instability was only in the internals. This still works for bools.

**Comment by simeonschaub** [comment]
> so the instability was only in the internals

I still see instabilities for `binomial`:

```julia
julia> Base.return_types(binomial, NTuple{2, Int16})
1-element Vector{Any}:
 Union{Int16, Int64}
```

> This still works for bools.

Can we also add a test to make sure this doesn't regress?

File: test/intfuncs.jl [hunk 1]
```diff
@@ -487,3 +487,15 @@ end
...
 end
+
+@testset "binomial" begin
+    for T in (Bool, Int8, UInt16, Int32, UInt64)
+        for x in rand(T, 1000)
+            @test binomial(x,1) == x
+            x < isqrt(typemax(T)) && @test binomial(x,2) == div(x*(x-1), 2)
+            @test binomial(x,x-1) == x
+            x < isqrt(typemax(T)) && @test binomial(x,x-2) == div(x*(x-1), 2)
+            @infered binomial(one(T),one(T))
```
Comment (dkarrasch) [inline test/intfuncs.jl:498]:
```suggestion
            @inferred binomial(one(T),one(T))
```

File: test/intfuncs.jl [hunk 2]
```diff
@@ -487,3 +487,15 @@ end
...
 end
+
+@testset "binomial" begin
+    for T in (Bool, Int8, UInt16, Int32, UInt64)
+        for x in rand(T, 1000)
+            @test binomial(x,T(1)) == x
+            x < isqrt(typemax(T)) && @test binomial(x,T(2)) == div(x*(x-1), 2)
+            @test binomial(x,T(x-1)) == x
+            x < isqrt(typemax(T)) && @test binomial(x,T(x-2)) == div(x*(x-1), 2)
+            @inferred binomial(one(T),one(T))
```
Comment (simeonschaub) [inline test/intfuncs.jl:498]:
```suggestion
            @inferred(binomial(one(T),one(T))) isa T
```

File: test/intfuncs.jl [hunk 3]
```diff
@@ -487,3 +487,15 @@ end
     ci = code_typed(() -> 14 // 21)[][1]
     @test ci.code == Any[Core.ReturnNode(2 // 3)]
 end
+
+@testset "binomial" begin
+    for T in (Bool, Int8, UInt16, Int32, UInt64)
+        for x in rand(T, 1000)
```
Comment (simeonschaub) [inline test/intfuncs.jl:493]: I think this is probably a little excessive 😂

Comment (oscardssmith) [reply test/intfuncs.jl]: It's still 100% compilation time...

Comment (simeonschaub) [reply test/intfuncs.jl]: I guess it won't hurt

**Review by simeonschaub — APPROVED** [review]

File: test/intfuncs.jl [hunk 4]
```diff
@@ -487,3 +487,17 @@ end
     ci = code_typed(() -> 14 // 21)[][1]
     @test ci.code == Any[Core.ReturnNode(2 // 3)]
 end
+@testset "binomial" begin
+    for T in (Int8, UInt16, Int32, UInt64)
+        for x in rand(isqrt(typemax(T)):isqrt(typemax(T)), 1000)
```
Comment (martinholters) [inline test/intfuncs.jl:492]: IIUC, `fill(isqrt(typemax(T)), 1000)` would be a more obvious way to put this, but the intention was probably something else?

Comment (oscardssmith) [reply test/intfuncs.jl]: oops.

File: test/intfuncs.jl [hunk 5]
```diff
@@ -487,3 +487,17 @@ end
     ci = code_typed(() -> 14 // 21)[][1]
     @test ci.code == Any[Core.ReturnNode(2 // 3)]
 end
+@testset "binomial" begin
+    for T in (Int8, UInt16, Int32, UInt64)
+        for x in rand(-isqrt(typemax(T)):isqrt(typemax(T)), 1000)
```
Comment (martinholters) [inline test/intfuncs.jl:492]: That doesn't work for the unsigned types I think. Maybe
```suggestion
        for x in rand(T(-isqrt(-widen(typemin(T)))):isqrt(typemax(T)), 1000)
```
would do it, but... uh...
Might be clearer to branch depending on whether `T` is signed or unsigned.

**Comment by oscardssmith** [comment]
Good to merge?

File: test/intfuncs.jl [hunk 6]
```diff
@@ -487,3 +487,17 @@ end
...
     @test ci.code == Any[Core.ReturnNode(2 // 3)]
 end
+@testset "binomial" begin
+    for T in (Int8, Int16, Int32, Int64)
+        for x in rand(-isqrt(typemax(T)):isqrt(typemax(T)), 1000)
+            @test binomial(x,T(1)) == x
+            x>=0 && @test binomial(x,x-T(1)) == x
+            @test binomial(x,T(2)) == div(x*(x-1), 2)
+            x>=0 && @test binomial(x,x-T(2)) == div(x*(x-1), 2)
+            @inferred(binomial(one(T),one(T))) isa T
```
Comment (dkarrasch) [inline test/intfuncs.jl:497]:
```suggestion
            @test @inferred(binomial(one(T),one(T))) isa T
```
?

Comment (oscardssmith) [reply test/intfuncs.jl:497]: good catch.

**Pull request merged**